Migrate publication

KludgeKML opened the pull request:
⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

## What

Allows frontend to render  items with schema_name `publication`

## Why

https://trello.com/c/utYFvGDW/781-move-publication-from-government-frontend-to-frontend, [Jira issue PNP-5828](https://gov-uk.atlassian.net/browse/PNP-5828)

## PR Review Page Examples

Live|Preview
------|------
https://www.gov.uk/government/publications/budget-2016-documents|https://govuk-frontend-app-pr-4927.herokuapp.com/government/publications/budget-2016-documents
https://www.gov.uk/government/statistics/women-in-northern-ireland-2017|https://govuk-frontend-app-pr-4927.herokuapp.com/government/statistics/women-in-northern-ireland-2017
https://www.gov.uk/government/statistics/uk-house-price-index-wales-june-2019.cy|https://govuk-frontend-app-pr-4927.herokuapp.com/government/statistics/uk-house-price-index-wales-june-2019.cy
https://www.gov.uk/government/publications/hs2-plan-and-profile-maps-2016-tibshelf-to-woodall-hsl13b|https://govuk-frontend-app-pr-4927.herokuapp.com/government/publications/hs2-plan-and-profile-maps-2016-tibshelf-to-woodall-hsl13b

## Screenshots

### https://www.gov.uk/government/statistics/women-in-northern-ireland-2017 and https://govuk-frontend-app-pr-4927.herokuapp.com/government/statistics/women-in-northern-ireland-2017

Before|After
------|------
<img width="2495" height="2471" alt="image" src="https://github.com/user-attachments/assets/65afe71d-bde3-4f02-bd29-079505b7881e" />|<img width="2495" height="2471" alt="image" src="https://github.com/user-attachments/assets/5077bf21-cd2d-404e-8985-426f031b436b" />

KludgeKML committed:
feat: Update README and examples config

Audit trail:
https://github.com/alphagov/government-frontend/blob/main/README.md?plain=1#L18

KludgeKML committed:
feat: Add placeholder route and controller for publications

- publications are all under either /government/publications or /government/statisticals/,
  but unfortunately they're not the only users of those routes, so we have to use
  a FullPathFormatRoutingConstraint to select them.
-there are examples in content store that are not in english, so path must support
  localisation

KludgeKML committed:
fix: alphabetise FullPathFormatRoutingConstraint options

KludgeKML committed:
tests: Add request tests

- At this point we have a basic controller / view / route, so the basic request tests we add for all migrated
  routes will pass.

KludgeKML committed:
feat: Copy publication view from government-frontend

- copies the view exactly as is from government-frontend. Subsequence commits will update the view to use values from the ContentItem model.

Audit trail:
https://github.com/alphagov/government-frontend/blob/aa70325b7750796c2bcfcb93ba5148d43956802a/app/views/content_items/publication.html.erb

KludgeKML committed:
lint: Fix lint errors in publication show template

- Run `bundle exec erb_lint --lint-all --autocorrect` to fix the errors

KludgeKML committed:
fix: replace @content_item in template with content_item

KludgeKML committed:
feat: create publication model

- and empty model spec

KludgeKML committed:
fix: Add meta description to document collection pages

- machine readable metadata tag only adds og:description, we have
  to add the default one.

KludgeKML committed:
fix: update heading values

- using the .name format begun in #4848

KludgeKML committed:
fix: add variable passed to translation partial

KludgeKML committed:
feat: make publication updatable

- view is about to require .updated and .initial_publication_date, provided by this concern.

KludgeKML committed:
feat: add SinglePageNotificationButton concern to publication

- needed for the next part of the template.
- uses concern added here: #4769

KludgeKML committed:
fix: simplify call to single_page_notification_button partial

KludgeKML committed:
feat: make publication political

- add Political concern and tests, this will be used by the shared/historical
  partial

KludgeKML committed:
fix: simplify call to history_notice partial

KludgeKML committed:
tests: add withdrawable test to model

- the next part of show we need to complete is the #withdrawn? call. Content items all contain Withdrawable
  by default, so we don't need to make any change to the model, but we have been checking each model
  for withdrawability with the shared behaviour.

KludgeKML committed:
fix: explicit variables in call to notice component

- Following the pattern in other migrations. This uses the methods from the automatically
  included Withdrawable concern.

KludgeKML committed:
feat: add national_applicability to publication

- Needed for the next part of the view

KludgeKML committed:
feat: copy localisation keys for publication

- run `bundle exec rails consolidation:copy_translation[publication,formats.publication]`

KludgeKML committed:
feat: add #datasets? model to publication

- this will be needed by the header data for the view
- add tests, as none existed in government-frontend

Audit trail:
https://github.com/alphagov/government-frontend/blob/8cb50cfaf715da5b5a99a4fd28db7f26b869e9c5/app/presenters/publication_presenter.rb#L32-L34

KludgeKML committed:
feat: Add PublicationPresenter

- next method to be added is clearly presentational, so add a presenter class and add it to the controller

KludgeKML committed:
feat: add #hide_from_search_engines? method to publication presenter

- this will be needed by the next view update.
- have checked with the app team and Jess which paths are still required, and have removed
  the others.
- simplified slightly by moving the exception into the constant array
- tests slightly reduced. Added risk is that someone updates the array in the presenter without
  updating the test, but this method is likely not to be valid for long, and since there are only
  two items,

Audit trail:
https://github.com/alphagov/government-frontend/blob/main/app/presenters/publication_presenter.rb#L9-L14
https://github.com/alphagov/government-frontend/blob/8cb50cfaf715da5b5a99a4fd28db7f26b869e9c5/app/presenters/publication_presenter.rb#L36-L41
https://github.com/alphagov/government-frontend/blob/8cb50cfaf715da5b5a99a4fd28db7f26b869e9c5/test/presenters/publication_presenter_test.rb#L113-L140

KludgeKML committed:
fix: Update headers in publication template

- following pattern from other route migrations
- we use the newly added methods

KludgeKML committed:
fix: fix test for attachments

- Initial shared test for `updates the preview_url base on asset_manager_id and maching filename`
  relied on the data supplied in the case_study example.

KludgeKML committed:
feat: add Attachments concern to Publication

- needed for the next part of the template

KludgeKML committed:
feat: add #featured_attachments Publication

- needed for the next part of the template
- this seems more model than strictly presentational, so we rename it to a
  more simple/model-like name to reflect what it's returning.

Audit trail:
https://github.com/alphagov/government-frontend/blob/f42d3c0775896474fae4584067382a88c2faa226/app/presenters/publication_presenter.rb#L20-L22
https://github.com/alphagov/government-frontend/blob/f42d3c0775896474fae4584067382a88c2faa226/test/presenters/publication_presenter_test.rb

KludgeKML committed:
feat: add #attachments_with_details_count

- it's debatable whether this is presentational or model, but we've previously added it to the
  model (https://github.com/alphagov/frontend/pull/4855), so keep the same call here. We
  might consider re-factoring how this is called in the presenter, but that requires changing
  some existing code so we'll make a card for that and keep things similar for now.

Audit trail:
https://github.com/alphagov/government-frontend/blob/f42d3c0775896474fae4584067382a88c2faa226/app/presenters/publication_presenter.rb#L24C1-L26C6

KludgeKML committed:
fix: use correct method

KludgeKML committed:
fix: update call

- following pattern in other migrations

KludgeKML committed:
test: Copy integration test for publication

- Copy the file across as is, changes in following commits.

Audit trail:
https://github.com/alphagov/government-frontend/blob/f42d3c0775896474fae4584067382a88c2faa226/test/integration/publication_test.rb

KludgeKML committed:
test: convert to empty spec test

SQUASH: convert to an empty spec test

KludgeKML committed:
fix: remove "random but valid items do not error" test

- this seems to have been added for a particular purpose,
  (https://github.com/alphagov/government-frontend/pull/294)
  and if it were needed in the future we should put it back as a
  request test.

KludgeKML committed:
test: convert title/description/body test

KludgeKML committed:
feat: add people support to publication model

- Will be needed in the metadata, as per:
  https://github.com/alphagov/government-frontend/blob/f42d3c0775896474fae4584067382a88c2faa226/test/integration/publication_test.rb#L22-L28

KludgeKML committed:
feat: add emphasised_organisations

- not explicitly tested in system tests, but the default for
  metadata in government-frontend is to include organisations
  sorted by emphasis, so we'll need this.

Audit trail:

https://github.com/alphagov/government-frontend/blob/f42d3c0775896474fae4584067382a88c2faa226/app/presenters/content_item/linkable.rb#L4

KludgeKML committed:
feat: render metadata

- following style for other migrations, use the publisher_metadata partial and explicitly pass
  values to it.
- uses updatable methods added in a previous commit
- uses contributor_links call from base content_item presenter, which calls the custom
  contributors method on the model.

KludgeKML committed:
feat: add #contributors method

- Publications need ordered orgs, plus people, since they use the default settings in
  government-frontend. The default also includes worldwide_organisations, but no
  publications in the content store have worldwide_organisations, and there's no
  way to add them in Whitehall, so we can ignore that.

Audit trail:
https://github.com/alphagov/government-frontend/blob/f42d3c0775896474fae4584067382a88c2faa226/app/presenters/content_item/linkable.rb#L4

KludgeKML committed:
test: does not render non-featured attachments

- we use the with-featured-attachments example from publishing-api,
  adding in another attachment which is not featured. (this context
  can be used for the following tests as well).

KludgeKML committed:
test: renders featured document attachments using components

KludgeKML committed:
fix: update partial name

KludgeKML committed:
fix: update translation keys

KludgeKML committed:
test: renders the accessible format option when the featured attachment is not accessible but an email is supplied

KludgeKML committed:
test: does not render the accessible format option when the featured attachment is when the featured attachment is not accessible and no email is supplied

KludgeKML committed:
test: does not render the accessible format option when the featured attachment is when the featured attachment is accessible and an email is supplied

- logic here is that if the document is already accessible, no need to render the request accessible option.

KludgeKML committed:
test: adds GA4 tracking to the accessible format option when the featured attachment is not accessible but an email is supplied

- This tracking is on the accessible format option, so we can add the test into the same context.

KludgeKML committed:
test: when attachments are external links, it renders them as external links, not files

- note that we change the expect here to PDF, not HTML, as that's what would actually be displayed here

KludgeKML committed:
fix: add title block to allow for withdrawn titles

- now that withdrawable is tested in the model and a presenter is available,
  add a title block that will show [WITHDRAWN] in the title.
- following pattern from other migrations

KludgeKML committed:
test: test exceptions for national statistics pages

KludgeKML committed:
test: withdrawn publication tests

- as used in other system tests

KludgeKML committed:
test: historical publication tests

- as used in other system tests

KludgeKML committed:
feat: Add national statistics logo images

KludgeKML committed:
feat: copy localisation keys for national_statistics logo alt text

- run `bundle exec rails consolidation:copy_translation[national_statistics,national_statistics]`
- note that the alt_text supports ky, but we don't
  have a logo for that. Worth considering if we want
  to display a default english logo.

Audit trail:
https://github.com/alphagov/government-frontend/tree/758d34770c63c572e0c94d9e4bde9c064b6a27c0/config/locales

KludgeKML committed:
feat: add NationalStatisticsLogo presenter concern

- this is going to be used in two places (publications and statistics_announcements) and
  was a presenter concern in government-frontend, so it makes sense to keep it that way
  here.
- we're a bit more explicit about what it does when the locale isn't supported (previously
  it would return a logo with no actual image, which would render an empty image with
  a placeholder text - except for ky, which has a translated text). Here we just don't
  return anything, but we might ask for a policy/content steer on that.
- we rely on the model having a #national_statistics? method. This is necessary because
  the two schemas that support it calculate it in different ways.

Audit trail:
https://github.com/alphagov/government-frontend/blob/aca90681e6b8bbb0aca7120f3027c604e110d53c/app/presenters/content_item/national_statistics_logo.rb

KludgeKML committed:
feat: add #national_statistics? method to Publication

- this is only used for presentation, but it is technically model info and having it here makes it a bit
  easier in the long run when statistics_announcements are added.

Audit trail:
https://github.com/alphagov/government-frontend/blob/main/app/presenters/publication_presenter.rb#L28-L30

KludgeKML committed:
feat: include NationalStatisticsLogo in PublicationPresenter

KludgeKML committed:
test: displays a devolved nations component with links to alternative URLs when visiting a publication that applies to a set of nations with alternative URLs

KludgeKML committed:
feat: renders with the single page notification button on English language pages

- note: we remove the exemption test here, as there are no exempt publications
  (of the four content_ids mentioned in government-frontend, two are guidance
  and two are no longer present).

Audit trail:
https://github.com/alphagov/government-frontend/blob/aca90681e6b8bbb0aca7120f3027c604e110d53c/app/presenters/content_item/single_page_notification_button.rb#L4-L9

KludgeKML committed:
test: does not render the single page notification button when the publication is in Welsh

KludgeKML committed:
test: returns the page with noindex meta tags when the path should be hidden from search engines

- moved to a request test because there's no argument that this is presented to the
  user in any way, so it's clearly in the request test domain.
- we don't need to test over the whole range because now there's only two examples,
  so if one of them is removed either it's not needed any more or we'll be alerted that
  the tests should be updated.

KludgeKML committed:
feat: test dates in Welsh publications

KludgeKML committed:
feat: update format names for statistics

- these were updated in government-frontend about a year ago.
- they'll be needed for the context values in the heading.

Audit trail:
- https://github.com/alphagov/government-frontend/pull/3177

deborahchua reviewed (approved):
This looks great! 👏 👏 Nicely ordered commits 🎉 
Left a comment about a typo but all good otherwise!

sairamya93 reviewed (commented):
Looking good.Nice trail of commits! 

Nitpick: Could we include the command run and the audit trail in this [commit](https://github.com/alphagov/frontend/pull/4927/commits/09891b85c0bbe26d3904deb16dcc13e99a54b129)?

KludgeKML committed:
test: convert metadata tests

- includes shared example for metatags
- use existing code from SchemaOrgHelpers in place of assertion.

Audit trail:
https://github.com/alphagov/government-frontend/blob/f42d3c0775896474fae4584067382a88c2faa226/test/test_helper.rb#L262-L271
